Added support for inline PTX assembly instructions.

MoFtZ opened the pull request:
Related to https://github.com/m4rs-mt/ILGPU/issues/416.

Provides a new `CudaAsm` class that can be used within a kernel to embed PTX instructions. This makes it easier to use advanced PTX instructions without the need to create a custom ILGPU intrinsic.

It supports multiple input arguments (arbitrarily capped at 10 in the T4 template), and a single output value per instruction statement.

```CSharp
public static void MyKernel(Index1D index, ArrayView<double> view)
{
    CudaAsm.Emit(
        "{\n\t" +
        "   .reg .f64 t1;\n\t" +
        "   add.f64 t1, %1, %2;\n\t" +
        "   add.f64 %0, t1, %2;\n\t" +
        "}",
        out double result,
        (double)index.X,
        42.0);
    view[index] = result;
}
```
The format string uses `%0` rather than the .NET way of `{0}`. This was to make it match the [inline PTX assembly](https://docs.nvidia.com/cuda/inline-ptx-assembly/index.html) in C++. If you want to use the `%`, then you should escape it with double `%%`.

UPDATE: The selective detection of PTX version, described below has been removed as it was causing issues. May revisit in a separate PR in the future.

It is also possible to selectively detect the PTX version available:
```CSharp
if (CudaAsm.IsPTXVersionAtLeast(CudaInstructionSet.ISA_70)) { }
else if (CudaAsm.IsPTXVersionAtLeast(CudaInstructionSet.ISA_60)) { }
else { }
```

KNOWN ISSUES:
- Using `CudaAsm.IsPTXVersionAtLeast` produces a complex constant expression that the ILGPU compiler is not currently able to fold into an unconditional branch. This means that the PTX compiler may fail if it does not support a newer instruction, that should have been optimised out.  Further work required to improve this area.

m4rs-mt commented on Src/ILGPU/Runtime/Cuda/CudaAsm.cs:
```diff
@@ -0,0 +1,60 @@
+        /// <returns>Return true if the PTX version is supported.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsPTXVersionAtLeast(int major, int minor) =>
+            IsSupported &&
```
I think if we change this `&&` implementation to a bitwise `$`, ILGPU should be able to fold this expression in all cases.

MoFtZ replied:
Unfortunately, using bitwise `&` does not help.

When `Accelerator.CurrentType` and `Accelerator.LanguageVersion` are specialised, they replace the placeholder IR node, but no folding it performed.

Using `IsSupported` as an example, the IR node will be `CompareValue(AcceleratorTypeValue, PrimitiveValue)`. When `Accelerator.CurrentType` is specialised, the `AcceleratorTypeValue` node is replaced by a `PrimitiveValue` - making the condition `CompareValue(PrimitiveValue, PrimitiveValue)` Later, when `UnreachableCodeElimination` is run, it finds that the condition is `CompareValue`, which cannot be folded - it only knows how to fold `PrimitiveValue`.

m4rs-mt replied:
You are right 😢 The aggressive constant-propagation pass in 57f3232228f2c4cea282570a300ffd92002fa412 should help to aggressively fold all expressions known at compile time. Perhaps right now is the right time to integrate this pass into ILGPU.

MoFtZ replied:
We originally discussed using the constant-propagation pass in the `O2` optimisation because it was expensive. However, after completing the work for `CudaAsm`, I realised that we need a solution for `O0` too. This is because in `O0` the PTX compiler will see all the PTX instructions, including future instructions that it does not know about.

m4rs-mt replied:
Summary of our discussion from the talk-to-dev session: we came up with the idea of limiting the scope of the (expensive) constant propagation pass to certain basic blocks. This helps to improve performance while being sufficient to optimize the expressions mentioned above.

m4rs-mt commented on Src/ILGPU/Backends/OpenCL/CLCodeGenerator.Values.cs:
```diff
@@ -706,5 +706,10 @@ public void GenerateCode(WarpShuffle shuffle)
+        /// <summary cref="IBackendCodeGenerator.GenerateCode(LanguageEmitValue)"/>
+        public void GenerateCode(LanguageEmitValue value) =>
+            // Ignore PTX instructions.
+            value.Assert(value.LanguageKind == LanguageKind.PTX);
```
I am not sure whether this assert is correct... 🤓

MoFtZ replied:
Correct. In future, if we add a new `LanguageKind`, this code will need to be modified to either ignore the new language, or emit the correct instructions.

m4rs-mt commented on Src/ILGPU/IR/Values/LanguageValues.cs:
```diff
@@ -0,0 +1,216 @@
+    /// Represents an inline lanaguage statement.
+    /// </summary>
+    [ValueKind(ValueKind.LanguageEmit)]
+    public sealed class LanguageEmitValue : Value
```
I guess we should make this class inherit from `MemoryValue` to avoid moving this value around due to possible side effects.

MoFtZ replied:
Done.

m4rs-mt commented on Src/ILGPU/IR/BasicBlock.cs:
```diff
@@ -453,8 +453,12 @@ public bool HasSideEffects()
-                if (value is MemoryValue || value is MethodCall)
+                if (value is MemoryValue ||
+                    value is MethodCall ||
+                    value is LanguageEmitValue)
```
If we change the base type to `MemoryValue` we can safely remove this again 🤓

MoFtZ replied:
Done.

m4rs-mt commented on Src/ILGPU/Frontend/Intrinsic/LanguageIntrinsics.cs:
```diff
@@ -0,0 +1,238 @@
+            // Valid all argument types
+            foreach (var arg in arguments)
+            {
+                if (arg.BasicValueType == BasicValueType.Int16 ||
```
May we add a comment here why `Int8` is not supported, since it cannot be represented properly in a non-existing 8-byte register?

MoFtZ replied:
This restriction is because the [inline PTX assembly](https://docs.nvidia.com/cuda/inline-ptx-assembly/index.html) did not have a register constraint for smaller types. However, on second thought, I think we should allow all primitive types and just deal with it in the PTX emitter.

m4rs-mt commented on Src/ILGPU/Frontend/Intrinsic/LanguageIntrinsics.cs:
```diff
@@ -0,0 +1,238 @@
+            while (expression.Length > 0)
+            {
+                // Search for next %
+                int foundIndex = expression.IndexOf('%');
```
Question: May we use a `Regex` instead?

MoFtZ replied:
This parsing scenario is easier than parsing `{0}`, so it could probably be done. I'll give it a try.

MoFtZ replied:
Done.

MoFtZ replied:
Added regex parsing of PTX expressions.

m4rs-mt commented on Src/ILGPU/IR/Transformations/AcceleratorSpecializer.cs:
```diff
@@ -58,6 +58,12 @@ public class AcceleratorSpecializer : SequentialUnorderedTransformation
+            /// <summary>
+            /// Returns the inline language version number for the  current accelerator.
+            /// </summary>
+            public readonly int AcceleratorLanguageVersion =>
```
Question: May we introduce an ILGPU-specific accelerator language-version enum?

m4rs-mt commented on Src/ILGPU/IR/Values/LanguageValues.cs:
```diff
@@ -0,0 +1,216 @@
+            VoidType voidType)
+            : base(initializer, voidType)
+        {
+            foreach (var argument in arguments)
```
Question: May we put this into `#if DEBUG` to avoid the emission of empty for-loop in release mode?

MoFtZ replied:
Fixed.

m4rs-mt commented on Src/ILGPU/Runtime/Accelerator.cs:
```diff
@@ -87,6 +87,30 @@ public static AcceleratorType CurrentType
+        /// <remarks>
+        /// The format is arbitrary, and assumes there are at most 99 minor versions.
+        /// </remarks>
+        internal static int MakeLanguageVersion(int major, int minor) =>
```
A got it... do you think an "arbitrary" version structure for ILGPU-specific inline language elements make sense (see comment about the `enumeration` above).

MoFtZ replied:
I was originally using `int MajorVersion` and `int MinorVersion`. This allowed easy matching in `IsPTXVersionAtLeast(major, minor)`. Later, I changed to a single `int LanguageVersion` so that the accelerator specialisation did not have to do two passes.

What would this enumeration look like? And how would we map `IsPTXVersionAtLeast(major, minor)` to this enumeration?

m4rs-mt replied:
Sounds good 👍 Maybe we can encapsulate the `int` in a parent structure that can be extended in the future without patching all uses of `int` 🤓

MoFtZ replied:
Changed to an opaque enum (`AcceleratorLanguageVersion`) to make future refactoring easier.

MoFtZ commented:
Replaced `int` language version with an opaque enum to make future refactoring easier.

Added support for all basic value types - previously did not support `Int1` and `Int8`. The user needs to take care since `Int1` maps to a predicate register, and `Int8` maps to a `Int16` register.

Changed PTX expression parsing to use regular expressions.

Changed the optimiser to deal with the known issue of removing PTX emit commands from branches that will never be reached. This was achieved by adding a new `PartialInlineMode` for performing partial IR rebuilding, and also changing the transformations used in `O0` mode. Also changed `&&` to `&` to avoid an issue with phi nodes that could not be folded, and therefore preventing the removal of some branches.

m4rs-mt commented on Src/ILGPU/IR/Construction/IRRebuilder.cs:
```diff
@@ -130,6 +131,111 @@ public interface IMode
+                foreach (var block in blocks)
+                {
+                    if (Predicate.Apply(block))
+                    {
```
nit: remove parenthesis 🤓

MoFtZ replied:
Done.

m4rs-mt commented on Src/ILGPU/IR/Transformations/AcceleratorSpecializer.cs:
```diff
@@ -158,8 +181,23 @@ public class AcceleratorSpecializer : SequentialUnorderedTransformation
+        /// Specializes accelerator inline language major version values.
+        /// </summary>
+        private static void Specialize(
+        RewriterContext context,
```
nit: Identation.

MoFtZ replied:
Done.

m4rs-mt commented on Src/ILGPU/IR/Transformations/AcceleratorSpecializer.cs:
```diff
@@ -382,7 +437,47 @@ static AcceleratorSpecializer()
+            // Merge the entry block
+            var entryBlock = builder.EntryBlockBuilder;
+            entryBlock.Clear();
+            entryBlock.MergeBlock(rebuilder.EntryBlock);
```
nice one, this is correct 👍

m4rs-mt commented on Src/ILGPU/IR/Transformations/Optimizer.cs:
```diff
@@ -164,16 +160,18 @@ public static class Optimizer
-            // could be created in prior passes
+            // Apply DCE phase to remove all dead values that could be created in prior
+            // passes
+            builder.Add(new UnreachableCodeElimination());
```
As we discussed, I fully agree that we need to apply these transformations in all cases. However, I still believe that the additional "DCE" pass in debug mode is not required. Am I right about that?

m4rs-mt commented on Src/ILGPU/IR/Transformations/AcceleratorSpecializer.cs:
```diff
@@ -382,7 +437,47 @@ static AcceleratorSpecializer()
+
+            // Create an exit return
+            var (exitBlock, exitValue) = rebuilder.Rebuild();
+            exitBlock.CreateReturn(exitValue.Location, exitValue);
```
I think this is correct since we are rebuilding all `affected` blocks. In other words, if we encounter an exit block, this block can perform a `return` operation only. All other blocks should have already been rewired to their successor blocks using  `Branch` terminator. Am I right?

MoFtZ replied:
Actually, I was not sure if this was always required. The original method should already have a an exit block and return statement, so is it really required? I copied this code from the constant propagation branch.

m4rs-mt commented on Src/ILGPU/IR/Construction/IRRebuilder.cs:
```diff
@@ -460,6 +637,31 @@ private void InitPhiValues(in BlockCollection blocks)
+                    var newNode = node.Value;
+
+                    if (oldNode != newNode &&
+                        oldNode.CanBeReplaced &&
```
Should we assert that this value can be replaced? Are there any other cases in which this can happen?

MoFtZ commented:
Removed code that supported selectively detecting the PTX version. May revisit in the future, possible when there is lambda support.
